src/chemistry/carma_aero/carma_clouds.F90

```diff
@@ -0,0 +1,528 @@
+
+    allocate( nspec(nbins) )
+
+    do m = 1, nbins
+       call rad_cnst_get_info_by_bin(0, m, nspec=nspec(m))
+    end do
+    ! add plus one to include number, total mmr and nspec
+    nspec_max = maxval(nspec)
+
+    ncnst_tot = nspec(1)
+    do m = 2, nbins
+      ncnst_tot = ncnst_tot + nspec(m)
+    end do
+
+   allocate(  bin_idx(nbins,nspec_max) )
+
+
+   ! Local indexing compresses the mode and number/mass indicies into one index.
+   ! This indexing is used by the pointer arrays used to reference state and pbuf
+   ! fields.
+   ! for CARMA we add number = 0, total mass = 1, and mass from each constituence into mm.
+   ii = 0
+   do m = 1, nbins
+      do l = 1, nspec(m)    ! loop through species
+         ii = ii + 1
+         bin_idx(m,l) = ii
+      end do
+   end do
+
+
+  end subroutine sox_cldaero_init
+
+!----------------------------------------------------------------------------------
+!----------------------------------------------------------------------------------
+  function sox_cldaero_create_obj(cldfrc, qcw, lwc, cfact, ncol, loffset) result( conc_obj )
+
+    real(r8), intent(in) :: cldfrc(:,:)
+    real(r8), intent(in) :: qcw(:,:,:)
+    real(r8), intent(in) :: lwc(:,:)
+    real(r8), intent(in) :: cfact(:,:)
+    integer,  intent(in) :: ncol
+    integer,  intent(in) :: loffset
+
+    real(r8) :: so4mmr(pcols,pver)
+    real(r8) :: nitmmr(pcols,pver)
+
+    type(cldaero_conc_t), pointer :: conc_obj
+
+    character(len=32) :: spectype
+
+    integer :: l,n,m
+    integer :: i,k,mm
+
+    ! local indexing for bins
+    !integer, allocatable :: bin_idx(:,:) ! table for local indexing of modal aero number and mmr
+
+
+    conc_obj => cldaero_allocate()
+
+    do k = 1,pver
+       do i = 1,ncol
+          if( cldfrc(i,k) >0._r8) then
+             conc_obj%xlwc(i,k) = lwc(i,k) *cfact(i,k) ! cloud water L(water)/L(air)
+             conc_obj%xlwc(i,k) = conc_obj%xlwc(i,k) / cldfrc(i,k) ! liquid water in the cloudy fraction of cell
+          else
+             conc_obj%xlwc(i,k) = 0._r8
+          endif
+       enddo
+    enddo
+
+    conc_obj%no3c(:,:) = 0._r8
+    conc_obj%nh4c(:,:) = 0._r8
+    conc_obj%so4c(:,:) = 0._r8
+
+    so4mmr(:,:) = 0._r8
+    do k = 1,pver
+       do i = 1,ncol
+          do m = 1, nbins
+            do l = 1, nspec(m)
+                  mm = bin_idx(m, l)
+                 call rad_cnst_get_bin_props_by_idx(0, m, l,spectype=spectype)
+                 if (trim(spectype) == 'sulfate') then
+                       so4mmr(i,k) =  so4mmr(i,k) +  qcw(i,k,mm)
+                 end if
+            end do
+          end do
+       end do
+    end do
+    conc_obj%so4c = so4mmr
+
+  end function sox_cldaero_create_obj
+
+
+!----------------------------------------------------------------------------------
+! Update the mixing ratios
+!----------------------------------------------------------------------------------
+  subroutine sox_cldaero_update( state, &
```

Comment:
Could split subroutine into sox_cldaero_tend to compute tendencies and sox_cldaero_update to apply tendencies.